docs/user-guide/provenance.ipynb

```diff
@@ -0,0 +1,279 @@
+   "metadata": {},
+   "source": [
+    "### Pipeline with a parameter table\n",
+    "\n",
+    "Pipelines with parameter tables can also be serialized but result in more complicated graphs."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "13d405df-0f15-4cdd-a275-4dcd3902ae7b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def to_string(x: float) -> str:\n",
+    "    return str(x)\n",
+    "\n",
+    "pipeline = sciline.Pipeline([to_string])\n",
+    "pipeline.set_param_table(sciline.ParamTable(int, {float: [3.0, -5.2]}))\n",
+    "task_graph = pipeline.get(sciline.Series[int, str])\n",
+    "task_graph.visualize(graph_attr={'rankdir': 'LR'})"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "41af9c26-6357-4db6-a495-8173027694df",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "show_json(task_graph.serialize())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5d8c0e23-3a06-485b-b90c-680ed4fa7a7a",
+   "metadata": {},
+   "source": [
+    "The serialized graph contains some nodes that are not shown in the visualization above.\n",
+    "Those nodes reveal some of the internals of `Pipeline` and `TaskGraph` and are needed to fully reconstruct the graph from JSON.\n",
+    "\n",
+    "- For the final `Series` object, there are two nodes: `\"Series[int, str]\"` and `\"provide_series[int, str]\"`.\n",
+    "  The former is a data node that represents the actual `Series` object, while the latter is a provider that assembles the series from its components.\n",
+    "- Some parameter table entries are also represented by two nodes: E.g., `\"float(int:0)\"` and `\"table_cell(float(int:0))\"`.\n",
+    "  The former is a special data node that contains properties which identify the table and location in the table that this data belongs to.\n",
+    "  The latter is a provider that produces said data from the user-defined table.\n",
+    "  Some properties of the former are required to fully identify the latter.\n",
+    "\n",
+    "In principle, the special provider nodes listed above could be omitted without losing information.\n",
+    "But they are included to ensure that, like `TaskGraph`, the serialized graph is bipartite in the sets of providers and data."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "279948ef-67ac-4434-9725-7b9b380c48ed",
+   "metadata": {},
+   "source": [
+    "### Reproducibility\n",
+    "\n",
+    "The JSON format used here was chosen as a simple, future-proof format for task graphs.\n",
+    "It can, however,  only capture part of the actual pipeline.\n",
+    "For example, it only shows the structure of the graph and contains the names of functions and types.\n",
+    "But it does not encode the implementation of those functions or types.\n",
+    "Thus, the graph can only be correctly reconstructed in an environment that contains the same software that was used to write the graph.\n",
```

Comment:
Would be nice to show an example of loading the json, building a new task graph from it and computing the result?
But I guess you've only implemented the saving of graphs, not loading them?

Reply:
I considered writing a loader. But that needs more info than is in the graph right now (parameters, package versions). And it's non-trivial because it needs some importlib finagling, so I wouldn't necessarily show it in the docs.